components/org.wso2.carbon.identity.tag.manager.core/src/main/java/org/wso2/carbon/identity/tag/manager/core/internal/impl/TagManagementServiceImpl.java

```diff
@@ -0,0 +1,271 @@
+        if (limit < 0) {
+            throw new TagServiceClientException(ErrorMessage.ERROR_CODE_INVALID_LIMIT_VALUE.getCode(),
+                    ErrorMessage.ERROR_CODE_INVALID_LIMIT_VALUE.getMessage());
+        }
+        if (offset < 0) {
+            throw new TagServiceClientException(ErrorMessage.ERROR_CODE_INVALID_OFFSET_VALUE.getCode(),
+                    ErrorMessage.ERROR_CODE_INVALID_OFFSET_VALUE.getMessage());
+        }
+        handleSorting(sortBy,sortOrder);
+        //TODO do filter validation
+        tagListResult = tagManagementDAO.filterTags(limit, offset, null, null, filter, tenantUuid);
+        return tagListResult;
+    }
+
+    @Override
+    public void updateTag(String tagUuid, String name, String description, boolean isPubliclyVisible)
+            throws TagServiceException {
+
+        if (StringUtils.isBlank(tagUuid)) {
+            throw new TagServiceClientException(ErrorMessage.ERROR_CODE_EMPTY_TAG_ID.getCode(),
+                    ErrorMessage.ERROR_CODE_EMPTY_TAG_ID.getMessage());
+        }
+        TagManagementDAO tagManagementDAO = DAOFactory.getInstance().getTagManagementDAO();
+        if (!tagManagementDAO.isExistingTagId(tagUuid)) {
+            throw new TagServiceClientException(ErrorMessage.ERROR_CODE_INVALID_TAG_UUID.getCode(),
+                    ErrorMessage.ERROR_CODE_INVALID_TAG_UUID.getMessage());
+        }
+        if (StringUtils.isBlank(name)) {
+            throw new TagServiceClientException(ErrorMessage.ERROR_CODE_EMPTY_NAME.getCode(),
+                    ErrorMessage.ERROR_CODE_EMPTY_NAME.getMessage());
+        }
+        tagManagementDAO.updateTag(tagUuid, name, description, isPubliclyVisible, tenantUuid);
+    }
+
+    @Override
+    public void deleteTag(String tagUuid) throws TagServiceException {
+
+        if (!isTagUUidValid(tagUuid)) {
+            throw new TagServiceClientException(ErrorMessage.ERROR_CODE_EMPTY_TAG_ID.getCode(),
+                    ErrorMessage.ERROR_CODE_EMPTY_TAG_ID.getMessage());
+        }
+        TagManagementDAO tagManagementDAO = DAOFactory.getInstance().getTagManagementDAO();
+        if (!tagManagementDAO.isExistingTagId(tagUuid)) {
+            throw new TagServiceClientException(ErrorMessage.ERROR_CODE_INVALID_TAG_UUID.getCode(),
+                    ErrorMessage.ERROR_CODE_INVALID_TAG_UUID.getMessage());
+        }
+        tagManagementDAO.deleteTag(tagUuid, tenantUuid);
+    }
+
+    @Override
+    public void addAssociation(String tagUuid, String resourceUuid) throws TagServiceException {
+
+        //TODO validate the resource type/uuid
+        TagManagementDAO tagManagementDAO = DAOFactory.getInstance().getTagManagementDAO();
+        if (StringUtils.isNotBlank(tagUuid) && StringUtils.isNotBlank(resourceUuid)) {
+            tagManagementDAO.storeAssociation(tagUuid, resourceUuid, tenantUuid);
+        } else {
+            throw new TagServiceClientException(ErrorMessage.ERROR_CODE_EMPTY_TAG_ID.getCode(),
```

Comment:
Why are we throwing `ERROR_CODE_EMPTY_TAG_ID` when the resourceUuid is blank?

Reply:
fixed